tests/cbmc/proofs/s2n_set_free_p/s2n_set_free_p_harness.c

```diff
@@ -21,12 +21,26 @@ void s2n_set_free_p_harness()
 {
     /* Non-deterministic inputs. */
     struct s2n_set *set = cbmc_allocate_s2n_set();
-    __CPROVER_assume(s2n_result_is_ok(s2n_set_validate(set)));
 
+    /* Assumptions. */
     nondet_s2n_mem_init();
+    __CPROVER_assume(s2n_result_is_ok(s2n_set_validate(set)));
 
     /* Operation under verification. */
-    if(s2n_result_is_ok(s2n_set_free_p(&set))) {
+    s2n_result result = s2n_set_free_p(&set);
+    if (s2n_result_is_ok(result)) {
         assert(set == NULL);
+    } else {
+        assert(s2n_errno != S2N_ERR_FREE_STATIC_BLOB);
     }
+
+    /* Cleanup after expected error cases, for memory leak check. */
+    if (s2n_result_is_error(result)) {
+        if (s2n_errno == S2N_ERR_NOT_INITIALIZED) {
+            /* failure due to s2n was not being initialized is expected. */
+            free(set->data->mem.data);
+            free(set->data);
+            free(set);
+        }
+   }
```

Comment:
I'd suggest you here (and in all the other harnesses with similar structure)
```c
    /* Operation under verification. */
    if (s2n_result_is_ok(s2n_set_free_p(&set))) {
        assert(set == NULL);
    } else {
        assert(s2n_errno != S2N_ERR_FREE_STATIC_BLOB);
        /* Cleanup after expected error cases, for memory leak check. */
        if (s2n_errno == S2N_ERR_NOT_INITIALIZED) {
            /* failure due to s2n was not being initialized is expected. */
            free(set->data->mem.data);
            free(set->data);
            free(set);
        }
   }
```

Reply:
I think it would be nice to keep the postconditions and the clean up part separate:
1. It makes debugging a lot simpler for non-memory-leak issues: comment the `--memory-leak-check` in `Makefile`, comment the cleanup code in harness, debug failing properties
2. We sometimes (although not in this particular harness) also need to cleanup if the function call was successful, so we would have to insert cleanup code in two different branches. I would rather insert two different branches in the cleanup code.

What do you think?

Reply:
(Feedback from offline discussion) You can add a comment instead explaining why we should separate these checks into different sections.

Reply:
Thanks, changed as discussed.